### DISCUSSION TITLE
Is there a way to store a user-defined collection metadata?

### DISCUSSION THREAD

**Discussion by r0x07k** [milvus-io/milvus#35356] [Q&A and General discussion] [answered]
Hi,

I would like to store some metadata for a collection, but it seems there is no convenient way to do so.

There is a 'properties' field that stores system properties, and Milvus accepts custom ones as well, but unfortunately, there is no Collection.get_properties function. I could store the metadata in a 'description' field, but it’s a somewhat makeshift solution.

So, it seems there is no robust way to store collection metadata apart from using an external database.

**Comment by xiaofan-luan** [comment]
so the metadata is for each collection?
I think it's easy to add some properties on collection, maybe you can add one. shouldn't be really hard

**Reply by r0x07k** [reply]
Yes, this is correct. For example, `"source":"30abf461-b78c-43ad-ad71-c08b60b318c8"`.

I can set it:
```
collection.set_properties(
    properties={
        "source":"30abf461-b78c-43ad-ad71-c08b60b318c8"
    }
)
```

But I cannot read it directly because there is no `collection.get_properties()`.

I could actually call `client.describe_collection()` and parse the whole json, so I will probably do that. But I think we do need `collection.get_properties()`. In both Python and Java.

**Comment by r0x07k** [comment]
So, we can store the metadata in 'properties' but only from Python.

In Python we can use the `Collection.set_properties` function to set the value and `client.describe_collection()['properties']` to get it.

In Java we can read:
```
            R<DescribeCollectionResponse> respDescribeCollection = milvusClient.describeCollection(
                DescribeCollectionParam.newBuilder()
                  .withCollectionName("")
                  .build()
              );
            respDescribeCollection.getData().getPropertiesList();
```

But cannot write :(. And this is the problem.

The reading in Java is also not that convenient. We can get the property only by index `public abstract E get(int index)`. To get it by name we will have to iterate through the list.

**Reply by yhmo** [reply, marked as answer]
In Java v2.4.2 or v2.3.8, you can add customized properties into a collection with alterCollection()

```
         milvusClient.alterCollection(AlterCollectionParam.newBuilder()
                .withCollectionName(COLLECTION_NAME)
                .withProperty("my_prop", "prop_value")
                .build());
        R<DescribeCollectionResponse> resp = milvusClient.describeCollection(DescribeCollectionParam.newBuilder()
                .withCollectionName(COLLECTION_NAME)
                .build());
        List<KeyValuePair> pairs = resp.getData().getPropertiesList();
        for (KeyValuePair pair : pairs) {
            System.out.println(pair);
        }
```

**Reply by r0x07k** [reply]
Thank you @yhmo .

Is it code for 2.3.8? Because for 2.4.2 the interface turned out to be completely different.

```
        Map<String, String> properties = Map.of("my_prop", "prop_value");
        client.alterCollection(AlterCollectionReq.builder()
                .collectionName(collectionName)
                .properties(properties)
                .build());
```

I couldn't confirm that it's working yet because it looks like Milvus replaced a `com.alibaba.fastjson.JSONObject` with a `com.google.gson.JsonObject` in the v2.4.2. So, a great deal of code will have to be rewritten to try this feature.

**Reply by yhmo** [reply]
A new version v2.4.3 of Java SDK has been released two days ago https://github.com/milvus-io/milvus-sdk-java/releases/tag/v2.4.3
There are two editions of MilvusClient in Java SDK.
1. MilvusClient --------------------- old client class
2. MilvusClientV2 ------------------ new client class for better/friendly interfaces

MilvusClient.alterCollection(AlterCollectionParam) is V1
MilvusClientV2.alterCollection(AlterCollectionReq) is V2

In v2.4.3, both classes can support set/get properties from collections.

**Reply by r0x07k** [reply]
It's clear now, thank you for explaining that.

**Comment by yiwen92** [comment]
@r0x07k Hi, I am curious why cannot store these info` "source":"30abf461-b78c-43ad-ad71-c08b60b318c8"` in collection's description? The metadata seems like a mapping info from your application, right?

**Comment by r0x07k** [comment]
Hi @yiwen92,

You're right. This and other fields we use link collections to external data sources and store application-specific flags and markers that are applied to the collection as a whole. The 'description' field in my understanding is intended for human-readable content where brief descriptions of the collection can be stored, while 'properties' is a precisely machine-readable dataset, parseable out of the box.
